MAINT/BUG: Don't use 5-sided quadrilaterals in Axes3D.plot_surface

eric-wieser opened the pull request:
Previously:
* "cell" perimeters were clumsily calculated with duplicates, which were then removed at runtime
* code to calculate normals was spread into multiple places
* average z was calculated even if not used
* repeated conversion between stride and count was done

Should have no visible behavior changes

This should be up to 25% faster, since the quadrilateral polygons now have 4 points, rather than 5 (!)

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+                    ztop   = a[rs,             cs:cs_next   ]
+                    zleft  = a[rs:rs_next,     cs_next      ]
+                    zbase  = a[rs_next,        cs_next:cs:-1]
+                    zright = a[rs_next:rs:-1,  cs           ]
```
This is the first time I've ever found myself using `a:b:-1` and it doing _exactly what I wanted_!

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+            for poly_i, ps in enumerate(polys):
+                # pick three points around the polygon to find the normal at
+                # hard to vectorize, since len(ps) is different at the edges
+                i1, i2, i3 = 0, int(len(ps)/3), int(2*len(ps)/3)
```
I feel like there's a much better approach here, but this is at least cheap, I suppose. Taking the middle point or even mean of each edge, and taking the cross product of the vectors joining opposite edges would be be a marginally better heuristic.

Either way, this is just cleanup, so I'm leaving it as is.

eric-wieser replied:
Note that if I want to make the image output unchanged, then this needs to become `(len(ps) + 1)`. Is per-pixel compatibility something that is cared about?

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+        row_inds = list(xrange(0, rows-1, rstride)) + [rows-1]
+        col_inds = list(xrange(0, cols-1, cstride)) + [cols-1]
 
 
```
Remove extra blank line.

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
 
 
-        #colset contains the data for coloring: either average z or the facecolor
+        #colset contains the sampled facecolors
```
Space after `#`.

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
-                    zbase = a[min(rows-1, rs+rstride), cs:min(cols, cs+cstride+1):][::-1]
-                    zright = a[rs:min(rows-1, rs+rstride):, cs][::-1]
+                    # the edges of the projected quadrilateral
+                    # note we use pythons half-open ranges to avoid repeating
```
Python's

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+                    # the edges of the projected quadrilateral
+                    # note we use pythons half-open ranges to avoid repeating
+                    # the corners
+                    ztop   = a[rs,             cs:cs_next   ]
```
Spacing is against PEP8.

eric-wieser replied:
Adds greatly to the clarity here though, so I'd maybe make an exception

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
-                avgzsum = sum(p[2] for p in ps2)
-                polys.append(ps2)
+                # ps = np.stack(ps, axis=-1)
+                ps = np.array(ps).T
```
I wonder if there's some way of pre-allocating this with the lengths of `row_inds` and `col_inds`.

eric-wieser replied:
No, because `ps` varies in length throughout the iteration

QuLogic replied:
Err, I might have had the wrong loop; the lengths of bits out of `X`, `Y`, Z` are not known?

eric-wieser replied:
No. They correspond the perimeters of the "square" "cells", which are not always the same size - the cells look like:

```
+---+---+-+
|   |   | |
|   |   | |
+---+---+-+
|   |   | |
|   |   | |
+---+---+-+
|   |   | |
+---+---+-+
```

where the trailing edges contain non-uniformly sized cells

QuLogic replied:
I think we're talking about different things; I mean `ps` itself which should be uniform since you're calling `np.array` on it, no?

eric-wieser replied:
Ah, I see. Yes, we could preallocate to `(2*(rs_next - rs) + 2*(cs_next - cs), 3)`

eric-wieser replied:
Ok, here's the answer - preallocating doesn't help, because `concatenate` has no `out` argument until numpy 1.14

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+                # pick three points around the polygon to find the normal at
+                # hard to vectorize, since len(ps) is different at the edges
+                i1, i2, i3 = 0, int(len(ps)/3), int(2*len(ps)/3)
+                v1[poly_i] = ps[i1] - ps[i2]
```
I prefer to be explicit with known 2D things, i.e., `v1[poly_i, :] = ps[i1, :] - ps[i2, :]`; I don't think there's any penalty for it.

eric-wieser replied:
There is a small penalty associated with parsing tuple indices over scalar indices, but I suppose it's probably not significant.

eric-wieser commented:
Looks like I got something wrong here...

eric-wieser commented:
Style fixes made, will investigate failures once I get my test setup working again

eric-wieser commented:
@QuLogic: Worked out why it was failing - previously, we were plotting every quadrilateral with 5 points, which changed the normal vector! I've updated the  images, and added another test that makes shading more obvious, and tests striding.

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1670,65 +1670,55 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
-                # are removed here.
-                ps = list(zip(*ps))
-                lastp = np.array([])
-                ps2 = [ps[0]] + [ps[i] for i in xrange(1, len(ps)) if ps[i] != ps[i-1]]
```
This didn't detect the fact that `ps[-1] == ps[0]`, always

eric-wieser commented:
Failures are PEP8 failures - one line that's 80 instead of 79, and then the extra spacing [here](https://github.com/matplotlib/matplotlib/pull/10001#discussion_r156877102)

eric-wieser commented:
Ping.

The failure is the following PEP8 violation:
```
/home/travis/build/matplotlib/matplotlib/lib/mpl_toolkits/mplot3d/axes3d.py:1686:25: E221 multiple spaces before operator
                    ztop   = a[rs,             cs:cs_next   ]
                        ^
/home/travis/build/matplotlib/matplotlib/lib/mpl_toolkits/mplot3d/axes3d.py:1686:60: E202 whitespace before ']'
                    ztop   = a[rs,             cs:cs_next   ]
                                                           ^
/home/travis/build/matplotlib/matplotlib/lib/mpl_toolkits/mplot3d/axes3d.py:1687:26: E221 multiple spaces before operator
                    zleft  = a[rs:rs_next,     cs_next      ]
                         ^
/home/travis/build/matplotlib/matplotlib/lib/mpl_toolkits/mplot3d/axes3d.py:1687:60: E202 whitespace before ']'
                    zleft  = a[rs:rs_next,     cs_next      ]
                                                           ^
/home/travis/build/matplotlib/matplotlib/lib/mpl_toolkits/mplot3d/axes3d.py:1688:26: E221 multiple spaces before operator
                    zbase  = a[rs_next,        cs_next:cs:-1]
                         ^
/home/travis/build/matplotlib/matplotlib/lib/mpl_toolkits/mplot3d/axes3d.py:1689:60: E202 whitespace before ']'
                    zright = a[rs_next:rs:-1,  cs           ]
```

1. Does this seem like a reasonable exception to spacing rules
2. Do I need to mark it as such?

anntzer commented:
I'm fine with putting a noqa.

eric-wieser commented:
Updated with:

* Flake8 ignore
* A helper function to enable a list comprehension

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1684,65 +1684,69 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
 
+        def boundary_edge(arr):
+            """
+            Get the boundary elements of the rectangle ``arr``,
```
Happy to rename / extract this function if someone else can suggest a name / location.

WeatherGod replied:
heh, I might call it "nibble()", because it nibbles around the edge of the 2D array. At the very least, I would expand the explanation a bit to get that point across. It took me a few minutes to realize that was what was done.

eric-wieser replied:
Not a fan of `nibble`. How about one of

[`array_`\](`perimeter`|`edge`|`boundary`|`border`)[`_`(`elements`|`values`)]

I think I like `array_perimeter` most.

Can you suggest a top-level place to put this in?

eric-wieser replied:
Moved to `cbook`, and added an example

WeatherGod commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1684,65 +1684,69 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+
+        # evenly spaced, and including both endpoints
+        row_inds = list(xrange(0, rows-1, rstride)) + [rows-1]
+        col_inds = list(xrange(0, cols-1, cstride)) + [cols-1]
```
this isn't py3k compatible. `xrange` doesn't exist, right?

WeatherGod replied:
Maybe `np.array_split()` would be cleaner? Don't know when it was added to numpy, but I do see it in v1.12.

eric-wieser replied:
`array_split` produces non-overlapping chunks, but an interesting idea.

I assume that there was `xrange = range` or something, but I guess not. I'll switch to `range`

eric-wieser replied:
It looks like I this came in after rebasing on #10525. Does matplotlib no longer support python 2?

QuLogic replied:
Not on `master`, it doesn't.

eric-wieser replied:
Am I right to target this against `master`?

QuLogic replied:
Yes, we can always backport if necessary (I don't think it is).

eric-wieser commented:
@WeatherGod: Updated, tests now all pass

eric-wieser commented:
@WeatherGod: Anything else needed here?

WeatherGod commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2027,3 +2066,4 @@ def _setattr_cm(obj, **kwargs):
                 delattr(obj, attr)
             else:
                 setattr(obj, attr, orig)
+
```
Looks like a PEP8 check is not liking the blank line here? I thought we were supposed to have a blank line at the end of files?

QuLogic replied:
A new line, not a full blank line.

WeatherGod replied:
Ah, I couldn't tell that with the github diffs. I like how git diff shows you those trailing blank lines.

eric-wieser replied:
Bad rebase I guess - will fix

WeatherGod commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1686,65 +1686,50 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
+                i1, i2, i3 = 0, len(ps)//3, 2*len(ps)//3
+                v1[poly_i, :] = ps[i1, :] - ps[i2, :]
+                v2[poly_i, :] = ps[i2, :] - ps[i3, :]
+            normals = np.cross(v1, v2)
```
We need an `else: normals = []` so that "normals" is always defined for the code blocks ahead.

eric-wieser replied:
I don't think this happens in any real cases, but will fix instead of trying to work out which cases.

Better would be a local function that's only called when the result is needed.

eric-wieser replied:
Setting to `[]` is an (existing) bug too, which I guess caused normals to be ignored when face colors were specified.

WeatherGod commented:
Besides those two points, I think this is ready to go. It is certainly easier to understand than before.

eric-wieser commented:
@WeatherGod: All fixed - thanks for catching that `normals` bug

WeatherGod commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -1681,71 +1681,60 @@ def plot_surface(self, X, Y, Z, *args, norm=None, vmin=None,
-        else :
-            normals = []
 
+        def get_normals(polygons):
```
fair enough

eric-wieser replied:
In particular, deciding which version to keep (assemble then cross vs cross then assemble) would require some profiling

eric-wieser commented:
Once this goes in, I'll pick back up #9990

eric-wieser commented:
@QuLogic, @anntzer, is there anything else you'd like to see here (that wouldn't fit better in a future PR)?

anntzer commented:
(I haven't actually followed the logic of the code, so I'll leave the review to others for now.)

eric-wieser commented:
Thanks! I'll get back to #9990 in the next few weeks
